week1/hw/part4/my_db_trans.erl

```diff
@@ -0,0 +1,74 @@
+%%%-------------------------------------------------------------------
+%%% @author Wojciech Pachuta
+%%%-------------------------------------------------------------------
+-module(my_db_trans).
+-author("Wojciech Pachuta").
+
+-define(SERVER, ?MODULE).
+
+%% API
+-export([start/0, stop/0, write/2, delete/1, read/1, match/1, lock/0, unlock/0]).
+
+%% Server callbacks
+-export([init/0, loop/1]).
+
+-record(state, {db, locking_pid}).
+
+start() -> spawn(?MODULE, init, []), ok.
+
+stop() -> ?SERVER ! stop, ok.
+
+write(Key, Element) -> ?SERVER ! {write, self(), {Key, Element}}, return_received().
+
+delete(Key) -> ?SERVER ! {delete, self(), Key}, return_received().
+
+read(Key) -> ?SERVER ! {read, self(), Key}, return_received().
+
+match(Element) -> ?SERVER ! {match, self(), Element}, return_received().
+
+%%server will unlock transaction when client dies
+%%its up to client to trap signals caused by server failure during transaction (between lock and unlock calls)
+lock() ->
+  ServerPid = whereis(?SERVER),
+  link(ServerPid),
+  ServerPid ! {lock, self()},
+  return_received().
+
+unlock() ->
+  ServerPid = whereis(?SERVER),
+  ServerPid ! {unlock, self()},
+  unlink(ServerPid),
+  return_received().
+
+
+init() ->
+  register(?SERVER, self()),
+  process_flag(trap_exit, true),
+  loop(#state{db = db:new(), locking_pid = none}).
+
+loop(#state{db = Db, locking_pid = LockingPid} = State) ->
+  receive
+    {'EXIT', LockingPid, _} -> loop(State#state{locking_pid = none});
```

Comment:
> What happens if a process that is not a client is linked to the server and dies?

`{'EXIT', DeadPid, Reason}` message will be left in mailbox and it wil never be accessed. But i cant just flush it, because this series of events would deadlock server:
1. client A calls lock()
2. server locks client A
3. client B calls lock()
4. client B dies
5. server flushes `{'EXIT', ClientBPid, Reason}`
6. client A unlocks
7. server unlocks client A
8. server locks client B, which is dead.
@mentels, how can i work around this problem?